src/main/java/org/terasology/projsndwv/systems/ApiarySystem.java

```diff
@@ -15,23 +15,273 @@
-        if (!entity.hasComponent(InventoryComponent.class)) {
-            InventoryComponent inventoryComponent = new InventoryComponent(9);
-            inventoryComponent.privateToOwner = false;
-            entity.addComponent(inventoryComponent);
+    public void onApiaryInventoryChanged(InventorySlotChangedEvent event, EntityRef entity, ApiaryComponent component) {
+        DelayManager delayManager = CoreRegistry.get(DelayManager.class);
+        if (delayManager == null) {
+            logger.error("No DelayManager in registry");
+            return;
+        }
+
+        if (event.getSlot() == SLOT_FEMALE) {
+            if (!event.getNewItem().hasComponent(BeeComponent.class)) {
+                entity.removeComponent(ApiaryMatingComponent.class);
+                if (delayManager.hasDelayedAction(entity, MATING_EVENT)) {
+                    delayManager.cancelDelayedAction(entity, MATING_EVENT);
+                }
+                if (delayManager.hasDelayedAction(entity, LIFE_TICK_EVENT)) {
+                    delayManager.cancelDelayedAction(entity, LIFE_TICK_EVENT);
+                }
+            }
+            else if (event.getNewItem().getComponent(BeeComponent.class).type == BeeComponent.BeeType.PRINCESS){
+                BeeComponent maleBee = entity.getComponent(InventoryComponent.class).itemSlots.get(SLOT_MALE).getComponent(BeeComponent.class);
+                if (maleBee != null) {
+                    entity.addOrSaveComponent(new ApiaryMatingComponent());
+                    delayManager.addDelayedAction(entity, MATING_EVENT, MATING_TIME);
+                }
+            }
+            else {
+                delayManager.addDelayedAction(entity, LIFE_TICK_EVENT, TempBeeRegistry.getTickTimeFromGenome(event.getNewItem().getComponent(GeneticsComponent.class).activeGenes.get(LOCUS_SPEED)));
+            }
+        }
+        else if (event.getSlot() == SLOT_MALE){
+            if (!event.getNewItem().hasComponent(BeeComponent.class)) {
+                entity.removeComponent(ApiaryMatingComponent.class);
+                if (delayManager.hasDelayedAction(entity, MATING_EVENT)) {
+                    delayManager.cancelDelayedAction(entity, MATING_EVENT);
+                }
+            }
+            else {
+                BeeComponent femaleBee = entity.getComponent(InventoryComponent.class).itemSlots.get(SLOT_FEMALE).getComponent(BeeComponent.class);
+                if (femaleBee != null) {
+                    if (femaleBee.type == BeeComponent.BeeType.PRINCESS) {
+                        entity.addOrSaveComponent(new ApiaryMatingComponent());
+                        delayManager.addDelayedAction(entity, MATING_EVENT, MATING_TIME);
+                    }
+                }
+            }
+        }
+    }
+
+    @ReceiveEvent
+    public void onApiaryEvent(DelayedActionTriggeredEvent event, EntityRef entity, ApiaryComponent component) {
```

Comment:
I would like to see this split up into smaller functions more. It's a bit too big and monolithic